README.md
Holy Mapperel
=============
An NES cartridge PCB manufacturing test by Damian Yerrick

This demo detects what mapper it's running on through mirroring,
tests how big of a ROM it was written on, verifies that the mapper
is mostly working, and displays the result.

The name
--------
Paul Molloy of infiniteneslives.com is selling INL-ROM circuit
boards for making NES game cartridges.  These boards are
configured at manufacturing time to support the mapper circuits
required by several games.  Among them are *Holy Diver* by IREM and
*Batman: Return of the Joker* by Sunsoft.  Putting those titles
together would produce something Burt Ward as Dick "Robin" Grayson
might say: "Holy diver, Batman!"

But the Warners probably wouldn't like that.  A label in Warner Music
Group publishes the Dio album after which *Holy Diver* was named, and
DC Comics (a unit of Warner Bros.) publishes comics starring Batman.
In 2010, DC forced NIS America to remove a Batman allusion from the
title of the PSP game now known as *What Did I Do To Deserve This,
My Lord?*.  So instead, I'm avoiding references to either Warner
company's works in the title.

Building
--------
First install Python 3, Pillow, cc65, GNU Make, and GNU Coreutils.
The README file for [nrom-template] describes how to set it up.

The makefile performs four steps:

1. Convert the system font from PNG format to an NES-friendly format
2. Assemble all source code files
3. Link source code files into mapperel-master.nes
4. Generate ROMs with several sizes and mappers

Because several mappers don't guarantee what bank will be visible
at power on, ROM generation repeats several parts of the program:

* Every 4 KiB has the Morse code crash handler.
* Every 16 KiB has a trampoline that attempts to change a discrete
  logic mapper or MMC1 to the last bank.
* Every 256 KiB has the entire test, in order to make testing of
  SUROM easier.

Once you have test ROMs of the appropriate size, you can use any
PRG/CHR splitting program.  The contents of all PRG ROMs of a
given size are identical no matter the mapper, as are all CHR
ROMs.  They differ only by header.  So go ahead and burn an EPROM
for each size and plug it into your socketed boards.  However, you
can't chop off the end of a larger ROM and use that on a smaller
ROM, as it will be missing the special tag that identifies it as
the last bank of the ROM.  This is intentional in order to prevent
builds with disconnected upper address lines from silently working.

[nrom-template]: https://github.com/pinobatch/nrom-template/

Supported boards
----------------
* 001: SxROM (Bionic Commando, Final Fantasy, Journey to Silius)
* 002: UNROM, UOROM (7432) (Battle Kid, Contra, DuckTales)
* 004: TxROM (Super Mario Bros. 2-3, Mega Man 3-6)
* 007: AxROM (Battletoads, Jeopardy!)
* 009: PNROM (Punch-Out!!)
* 010: FJROM, FKROM (Famicom Wars, Fire Emblem)
* 011: Color Dreams (Crystal Mines, Spiritual Warfare)
* 028: INL-ROM (A53) (Action 53 volumes 1 and 2)
* 034: BNROM (Deadly Towers, Haunted: Halloween '85)
* 066: NROM, CNROM, GNROM (Balloon Fight, Pipe Dream, Gumshoe)
* 069: JLROM, JSROM (Batman: Return of the Joker, Gimmick!, Hebereke)
* 078.3: IF-12 (Holy Diver)
* 118 TxSROM (NES Play Action Football, Wanderers from Ys)
* 180: UNROM (7408) (Crazy Climber)

The tests
---------
The tests look at bank tags, which are increasing numbers spaced at
regular intervals in memory that let the program know what bank has
been switched into a particular address window.  Bank tags are placed
at offset 4088 ($FF8) of every 4 KiB bank of PRG ROM and at offset
508 ($1FC) of every 2 KiB bank of CHR ROM or RAM.

First it copies test code into RAM that detects the mapper by how
it responds to writes to the supported boards' nametable mirroring
ports.  Then, because a few mappers react the same way, such as all
boards with fixed mirroring, it narrows the field through very basic
bank switching tests.  This needs to be done in RAM because 32K bank
switching may cause the startup ROM bank to be switched out.

After the mapper number is determined, it copies a driver for that
mapper into RAM.  The driver is responsible for changing 8K CHR banks
and WRAM banks and doing detailed tests.

### CHR test

The CHR test starts by asking the driver to switch the first 8K of
CHR memory into PPU $0000-$1FFF.  If data written to $0000 in CHR
space does not change, it's CHR ROM.  Otherwise, it's CHR RAM, and
the test writes bank tags in CHR RAM banks 31 down to 0.  It reads
bank tags from the last CHR bank to verify that they're internally
consistent.  If so, it knows the size of CHR memory.

If CHR ROM is present, the test coompares the font in $0000-$03FF to
a copy of the font in PRG ROM to ensure that CHR ROM can be read.
Then it verifies all bank tags and saves the result for later.

If it's CHR RAM, it writes a pseudorandom pattern to all of CHR RAM,
reads it back, and saves whether it matched for later.  Then it
repeats the test seven more times with the pattern shifted by a byte
each time.  While this is going on, it buzzes the speaker to let the
user know it hasn't frozen.  Finally, it puts the bank tags back and
loads the small font into CHR RAM.

### WRAM test 

Work RAM (WRAM), or PRG RAM, is RAM on the cartridge at CPU
$6000-$7FFF.  First it looks for the string `SAVEDATA` at $6100 and
sets a flag if it was found.  Then the same test is done as for CHR
RAM:  determine how much, then test each byte.  Finally, `SAVEDATA`
is written back.  This way, the user can test for battery backup by
running the test, powering off, and running the test again.
Pressing the Reset button will cause an incorrect result.

### Detailed test

Some mappers are more flexible than others in how they map PRG and
CHR banks into windows within the CPU and PPU address space.  This
test steps through all the PRG and CHR bank numbers in each window
with various combinations of banking modes.  It also checks whether
WRAM can be disabled for power-off protection (on mappers that claim
to support this) and whether the IRQ works roughly as expected.
This is no substitute for an exhaustive mapper-specific test, but
it should help determine whether the chips are soldered properly.

Morse codes
-----------
Sometimes a test fails hard enough that the framework isn't even sure
that it can display the results on screen.  When a test fails this
hard, your NES will beep Morse code at you.  Find a ham and fix your
soldering.

* `WB` (`.__ _...`):  
  Wrong bank at startup.  INL's versions of the ASIC mappers
  guarantee that the LAST 4 KiB of the cart is swapped into
  $F000-$FFFF at power on.  Discrete mappers should be OK as well,
  as there's a stub every 16K.
* `MIR` (`__ .. ._.`):  
  The nametable mirroring for this mapper doesn't match any of the
  supported mappers.  Check PA13-PA10, /PA13, CIRAM A10, and CIRAM
  enable, and don't try running the 78.3 test on an emulator that
  does not support NES 2.0 format.
* `SU` (`... .._`):  
  Attempt to switch to second half of SUROM (4 Mbit MMC1) failed.
* `LB` (`._.. _...`):  
  Mapper detection did not leave the last 16K of PRG ROM swapped in.
* `RB` (`._. _...`):  
  Attempt to return to the last bank after a test failed.
* `CBT` (`_._. _... _`):  
  CHR bank tags are not internally consistent.
* `FON` (`.._. ___ _.`):  
  Font in CHR memory does not match that in PRG ROM.  The CHR ROM or
  CHR RAM is bad.
* `DRV` (`_.. ._. ..._`):  
  The mapper was detected, but no PRG and CHR test driver for the
  mapper exists.
* `SMS` (`... __ ...`):  
  Your mobile phone has received a text message.

The mapper 34 test may freeze in emulators that do not support the
NES 2.0 format.  They make both the BNROM registers and the NINA-001
registers available at the same time, when the current best practice
is to switch between the two based on CHR size.

Displayed result
----------------
After the buzzing ends, the following variables are valid:

* `cur_mapper`: iNES mapper number of the detected mapper
* `last_prg_bank`: Size of PRG ROM in 4096 byte units, minus one
* `is_chrrom`: Zero if CHR RAM is present; nonzero if CHR ROM is
  present
* `last_chr_bank`: Size of CHR ROM or CHR RAM in 8192 byte units,
  minus one
* `chr_test_result`: Zero if bank tags were read correctly from CHR
  ROM or if all bytes of CHR RAM correctly hold values.
* `has_wram`: Nonzero if $6000-$7FFF is RAM.
* `has_savedata`: Nonzero if `SAVEDATA` was in $6100 of the first
  WRAM bank at startup
* `last_wram_bank`: WRAM size in 8192 byte units, minus one;
  unspecified if `has_wram` is zero
* `wram_test_result`: Zero if all bytes of WRAM correctly hold
  values; unspecified if `has_wram` is zero
* `driver_prg_result`: Detailed WRAM and PRG ROM test results
* `driver_chr_result` Detailed IRQ and CHR test results

All this information is displayed on the screen.  The detailed code
is 4 digits: WRAM, PRG ROM, IRQ, and CHR ROM/RAM.  Zero is normal;
anything else reflects something unexpected.

* Mapper 001 `1xxx`: $E000 bit 4 does not disable WRAM
* Mapper 001 `4xxx`: $A000 bit 4 does not disable WRAM (SNROM), or
  $A000 bit 4 disables WRAM (all but SNROM)
* Mapper 004 `2xxx`: Read-only mode not present

In iNES format environments that don't support NES 2.0, such as FCEUX
and PowerPak, the MMC3 test will return a warning about lack of write
protection on WRAM.  These environments don't implement this feature
because not implementing it gets the two *StarTropics* games, which
use MMC6, to run with the MMC3 driver.

Some INL-ROM products implement only a subset of the entire mapper as
a cost-saving measure.  For example, MMC3 may lack WRAM protection or
may fix C and P values.  Detailed codes will reflect unimplemented
features.  If a subset changes mirroring behavior, the mapper may not
be recognized in the first place.

For the convenience of users testing large numbers of boards without
looking at the screen, the result is also beeped through the speaker
with one note per nibble so that an attentive ear can pick up any
deviation from the intended melody.

    0: Bb       4: F        8: Mid C    C: G
    1: Low C    5: G        9: D        D: A
    2: D        6: A        A: E        E: B
    3: Eb       7: Bb       B: F        F: High C

There are three groups of nibbles:

1. Tweet, mapper number (2)
2. Tweet, PRG ROM size in 32768 byte units (2), PRG RAM size in
   8192 byte units (2), buzz if PRG RAM defective, ding if battery,
   PRG RAM detailed result (1), PRG ROM detailed result (1)
3. Tweet, CHR size in 8192 byte units minus 1 (2), CHR RAM or
   ROM (1), buzz if CHR RAM or ROM defective, IRQ detailed
   result (1), CHR detailed result (1)

Limits
------
Not all mappers are supported.

NROM and CNROM are detected as mapper 66, but the board name should
be correct if their memory size matches a known configuration.

The IRQ test measures only gross behavior of the interval timer of
the MMC3 or FME-7.  It is intended to measure continuity of the IRQ
pin during board assembly, not as a substitute for detailed tests of
exact timing of a CPLD reimplementation of a particular mapper.

The "Flash ID" results mean nothing.  They are reserved for
future expansion.  Pull requests are appreciated.

Legal
-----
Copyright 2013-2017 Damian Yerrick

Available under zlib License.

Not sponsored or endorsed by DC Comics, IREM, Nintendo, or Sunsoft.


CHANGES.txt
0.03 (future)
* Port tools to Python 3.8 or later (collections.abc)

0.02 (2017-10-16)
* Rename to "Holy Mapperel" to avoid Warner Bros. trouble
* Port tools to Python 3
* Port asm and makefile to ca65 2.14 or later (requested by
  infiniteneslives)
* make_roms.py: Generate 2K, 8K, and 32K save files
* README: List dits and dahs (requested by rainwarrior)
* README: Clarify what's duplicated how often, why Morse crashes are
  used, and why the end of a ROM can't just be cut off (requested
  by rainwarrior)
* Relicense from GNU All-Permissive to the equivalent (but more
  widely recognized) zlib
* Add Color Dreams mapper
* Distinguish mappers 0 (NROM) and 3 (CNROM) from their superset
  66 (GNROM), and 34 (BNROM) from 11 (Color Dreams)

0.01 (2013-11-02)
* Initial public release


tools/make_roms.py
#!/usr/bin/env python3
"""
ROM builder for Holy Mapperel

Copyright 2013, 2022 Damian Yerrick

This software is provided 'as-is', without any express or implied warranty. In
no event will the authors be held liable for any damages arising from the use of
this software.

Permission is granted to anyone to use this software for any purpose, including
commercial applications, and to alter it and redistribute it freely, subject to
the following restrictions:

1.  The origin of this software must not be misrepresented; you must not claim
    that you wrote the original software. If you use this software in a product,
    an acknowledgment in the product documentation would be appreciated but is
    not required.

2.  Altered source versions must be plainly marked as such, and must not be
    misrepresented as being the original software.

3.  This notice may not be removed or altered from any source distribution.
"""
INES_MIRRH = 0
INES_MIRRV = 1
INES_MIRR4 = 8
INES_NTSC = 0
INES_PAL = 1
INES_DUAL_TVSYSTEM = 2
MAPPER_NROM = 0
MAPPER_MMC1 = 1
MAPPER_UNROM = 2
MAPPER_CNROM = 3
MAPPER_MMC3 = 4
MAPPER_MMC5 = 5
MAPPER_AOROM = 7
MAPPER_MMC2 = 9
MAPPER_MMC4 = 10
MAPPER_COLORDREAMS = 11
MAPPER_CPROM = 13
MAPPER_A53 = 28
MAPPER_BNROM = 34
MAPPER_GNROM = 66
MAPPER_FME7 = 69
MAPPER_HOLYDIVER = (78, 3)
MAPPER_MMC3_TLSROM = 118
MAPPER_MMC3_TQROM = 119
MAPPER_UNROM_CRAZY = 180
MAPPER_ACTIVE = 228

def log2(value):
    if value < 1:
        return -1
    logvalue = 0
    while value >= 0x1000000:
        value >>= 24
        logvalue += 24
    if value >= 0x10000:
        value >>= 16
        logvalue += 16
    if value >= 0x100:
        value >>= 8
        logvalue += 8
    if value >= 0x10:
        value >>= 4
        logvalue += 4
    if value >= 4:
        value >>= 2
        logvalue += 2
    return logvalue + (value >> 1)

def make_nes2_ramsize(size):
    if not 0 <= size <= 1048576:
        return ValueError("RAM size must be 0 to 1048576 bytes, not %d" % size)
    if size == 0:
        return 0
    return max(1, log2(size - 1) - 5)
    
def make_nes2_ramsizes(unbacked, backed):
    unbacked = make_nes2_ramsize(unbacked)
    backed = make_nes2_ramsize(backed)
    return unbacked | (backed << 4)

valid_mirrorings = {
    0: 0, 1: 1, 8: 8, 9: 9,
    'H': 0, 'V': 1, 'h': 0, 'v': 1, '4': 8
}

def format_memsize(size):
    if size < 1024:
        return "%d" % size
    if size < 1048576:
        return "%dK" % (size // 1024)
    return "%dM" % (size // 1048576)

def make_nes2_header(prgsize, chrsize=0, mapper=0, mirroring=INES_MIRRV,
                     prgramsize=0, chrramsize=0, tvsystem=0):
    """Make a byte string representing a 16-byte NES 2.0 header.

prgsize -- Size of PRG ROM in bytes (multiple of 16384)
chrsize -- Size of CHR ROM in bytes (multiple of 8192)
mapper -- iNES mapper number (0-4095) or a tuple (mapper, submapper)
mirroring -- iNES mirroring code (0, 1, 8, 9) or letter ('H', 'V', '4')
prgramsize -- Sizes of PRG RAM as tuple (not battery-backed, battery-backed)
chrramsize -- Sizes of CHR RAM as tuple (not battery-backed, battery-backed)

"""
    if isinstance(mapper, tuple):
        mapper, submapper = mapper
    else:
        submapper = 0
    if not isinstance(prgramsize, tuple):
        prgramsize = (prgramsize, 0)
    if not isinstance(chrramsize, tuple):
        chrramsize = (chrramsize, 0)

    if not 16384 <= prgsize < 4096 * 16384:
        raise ValueError("PRG ROM size must be 16384 to 67092480 bytes, not %d" % prgsize)
    if prgsize % 16384:
        raise ValueError("PRG ROM size must be a multiple of 16384 bytes, not %d" % prgsize)
    prgsize = prgsize // 16384
    if not 0 <= chrsize < 4096 * 8192:
        raise ValueError("CHR ROM size must be 0 to 33546240 bytes, not %d" % chrsize)
    if chrsize % 8192:
        raise ValueError("CHR ROM size must be a multiple of 8192 bytes, not %d" % chrsize)
    chrsize = chrsize // 8192
    if not 0 <= mapper < 4096:
        raise ValueError("mapper must be 0 to 4095, not %d" % mapper)
    if not 0 <= submapper < 16:
        raise ValueError("submapper must be 0 to 15, not %d" % submapper)
    try:
        mirroring = valid_mirrorings[mirroring]
    except KeyError:
        raise ValueError("mirroring must be 0, 1, 8, 9, 'H', or 'V', not %s" % mirroring)
    if tvsystem >= 4:
        raise ValueError("mirroring must be 0-3, not %d" % tvsystem)
    prgramsize = make_nes2_ramsizes(*prgramsize)
    chrramsize = make_nes2_ramsizes(*chrramsize)
    battery = 2 if ((chrramsize | prgramsize) & 0xF0) else 0

    header = bytearray(b"NES\x1a")
    header.append(prgsize & 0x0FF)
    header.append(chrsize & 0x0FF)
    header.append(mirroring | battery | ((mapper & 0x00F) << 4))
    header.append((mapper & 0x0F0) | 0x08)

    header.append(((mapper & 0xF00) >> 8) | (submapper << 4))
    header.append(((prgsize & 0xF00) >> 8) | ((chrsize & 0xF00) >> 4))
    header.append(prgramsize)
    header.append(chrramsize)
    header.append(tvsystem)
    header.extend([0] * (16 - len(header)))
    return bytes(header)

# PRG ROM size or range, CHR ROM size or range,
# mapper, mirroring, PRG RAM size(s), CHR RAM sizes
romspecs_all = [
    # Discretes
    (32768, 8192,
     MAPPER_NROM, (INES_MIRRV, INES_MIRRH), (0, 2048, 4096), 0),
    (32768, (16384, 32768),
     MAPPER_CNROM, (INES_MIRRV, INES_MIRRH), 0, 0),
    (32768, 0,
     MAPPER_CNROM, (INES_MIRRV, INES_MIRRH), 0, 32768),
    ((32768, 524288), 0,
     MAPPER_UNROM, (INES_MIRRV, INES_MIRRH), 0, 8192),
    ((32768, 524288), 0,
     MAPPER_UNROM_CRAZY, (INES_MIRRV, INES_MIRRH), 0, 8192),
    (32768, 0,
     MAPPER_CPROM, (INES_MIRRV, INES_MIRRH), 0, 16384),
    ((32768, 262144), 0,
     MAPPER_AOROM, 0, 0, 8192),
    ((32768, 524288), 0,
     MAPPER_BNROM, (INES_MIRRV, INES_MIRRH), 0, 8192),
    ((65536, 131072), (16384, 32768),
     MAPPER_GNROM, (INES_MIRRV, INES_MIRRH), 0, 0),
    ((65536, 131072), (16384, 131072),
     MAPPER_COLORDREAMS, (INES_MIRRV, INES_MIRRH), 0, 0),
    # use of Color Dreams to provide expanded CHR RAM for BNROM
    ((65536, 131072), 0,
     MAPPER_COLORDREAMS, (INES_MIRRV, INES_MIRRH), 0, (8192, 32768)),

    # SGROM SNROM SUROM SOROM SXROM
    ((32768, 524288), 0,
     MAPPER_MMC1, 0, (0, 8192, (8192, 8192), 32768), 8192),
    # MMC1 with CHR ROM
    ((32768, 262144), (16384, 131072),
     MAPPER_MMC1, 0, (0, 8192), 0),
    # TKSROM TLSROM
    ((32768, 524288), (16384, 262144),
     MAPPER_MMC3_TLSROM, 0, (0, 8192), 0),
    # Mega Man 4/6, Final Fantasy III, Full Quiet
    ((32768, 524288), 0,
     MAPPER_MMC3, 0, (0, 8192), (8192, 32768)),
    # Rest of MMC3
    ((32768, 524288), (16384, 262144),
     MAPPER_MMC3, 0, (0, 8192), 0),
    # BTR and JxROM
    ((32768, 262144), (16384, 262144),
     MAPPER_FME7, 0, (0, 8192), 0),
    # Holy Diver
    ((32768, 131072), (16384, 131072),
     MAPPER_HOLYDIVER, 0, 0, 0),
    # PNROM
    ((32768, 131072), (16384, 131072),
     MAPPER_MMC2, 0, 0, 0),
    # FxROM
    ((32768, 131072), (16384, 131072),
     MAPPER_MMC4, 0, (0, 8192), 0),
]

romspecs_oneofeach = [
    #   PRG     CHR  Mapper              Mirror, WRAM, CHR RAM
    ( 32768,   8192, MAPPER_NROM,        INES_MIRRV, 0, 0),
    ( 32768,      0, MAPPER_NROM,        INES_MIRRV, 0, 8192),
    ( 32768,  32768, MAPPER_CNROM,       INES_MIRRH, 0, 0),
    ( 32768,      0, MAPPER_NROM,        INES_MIRRV, 0, 32768),
    (131072,      0, MAPPER_UNROM,       INES_MIRRV, 0, 8192),
    (131072,      0, MAPPER_UNROM_CRAZY, INES_MIRRH, 0, 8192),
    (131072,      0, MAPPER_AOROM,       0, 0, 8192),
    (131072,      0, MAPPER_BNROM,       INES_MIRRH, 0, 8192),
    ( 65536,  65536, MAPPER_COLORDREAMS, INES_MIRRV, 0, 0),
    ( 65536,      0, MAPPER_COLORDREAMS, INES_MIRRV, 0, 32768),
    ( 65536,  16384, MAPPER_GNROM,       INES_MIRRV, 0, 0),
    (131072,  32768, MAPPER_MMC1,        0, (0, 8192), 0),
    (131072,      0, MAPPER_MMC1,        0, 0, 8192),
    (524288,      0, MAPPER_MMC1,        0, ((0, 8192), (0, 32768)), 8192),
    (131072, 131072, MAPPER_MMC1,        0, (0, 8192), 0),
    (131072,  65536, MAPPER_MMC2,        INES_MIRRV, 0, 0),
    (262144, 262144, MAPPER_MMC3,        0, 0, 0),
    (262144,      0, MAPPER_MMC3,        0, 0, 32768),
    ( 1<<20,      0, MAPPER_MMC3,        0, 0, 32768),  # Oversize; HB only
    (131072,      0, MAPPER_MMC3,        0, 0, (8192, 32768)),
    (131072,  65536, MAPPER_MMC3_TLSROM, 0, 0, 0),
    (131072,  65536, MAPPER_MMC4,        INES_MIRRV, 8192, 0),
    (131072,  65536, MAPPER_FME7,        0, 8192, 0),
    (131072,  65536, MAPPER_HOLYDIVER,   0, 0, 0),
    (131072,      0, MAPPER_HOLYDIVER,   0, 0, 32768),
    (524288,      0, MAPPER_A53,         0, 0, 32768),
    ( 1<<20,      0, MAPPER_A53,         0, 0, 32768),
]

romspecs = romspecs_oneofeach

def log_xrange(start=1, end=None, step=2):
    if end is None:
        start, end = 1, start
    while start <= end:
        yield start
        start = start * step

filename_mirroring = {0: 'H', 1: 'V', 8: '4', 9: '4V'}
switchable_mirror_mappers = {
    MAPPER_MMC1, MAPPER_MMC2, MAPPER_MMC3, MAPPER_MMC3_TQROM,
    MAPPER_MMC3_TLSROM, MAPPER_MMC4, MAPPER_AOROM, MAPPER_HOLYDIVER,
    MAPPER_FME7, MAPPER_A53
}

def handle_single_rom(prgsize, chrsize, mapper, mirror,
                      prgramsize, chrramsize):
    filename = ['M%d' % (mapper[0] if isinstance(mapper, tuple) else mapper),
                '.%d' % mapper[1] if isinstance(mapper, tuple) else '',
                '_P', format_memsize(prgsize),
                '_C%s' % format_memsize(chrsize) if chrsize else '',
                '_CR%s' % format_memsize(chrramsize) if chrramsize else '',
                '_%s' % filename_mirroring[mirror & 0x09]
                if mapper not in switchable_mirror_mappers
                else '',
                '_W%s' % format_memsize(prgramsize[0]) if prgramsize[0] else '',
                '_S%s' % format_memsize(prgramsize[1]) if prgramsize[1] else '',
                '.nes']
    filename = ''.join(filename)
##    chrramsize = (8192 if mapper == MAPPER_MMC3_TQROM
##                  else 0 if chrsize > 0
##                  else 16384 if mapper == MAPPER_CPROM
##                  else 8192)
    header = make_nes2_header(prgsize, chrsize, mapper, mirror,
                              prgramsize, chrramsize)

    # SUROM/SXROM can't guarantee PRG A18 until CHR is set up
    # so duplicate the test in all 256K outer banks
    dupli_prgsize = min(262144, prgsize)

    # Place right bank in the last bank
    prgrom = bytearray([0xFF]) * (dupli_prgsize - 32768)
    prgrom.extend(primary_prgrom)

    # Place wrong bank code into all 4K banks
    wrong_bank = bytearray(primary_prgrom[0x7F80:0x8000])
    wrong_bank[-4] = 0x80  # Set reset vector
    wrong_bank[-3] = 0xFF
    for i in range(4096, dupli_prgsize, 4096):
        prgrom[i - 128:i] = wrong_bank
    del wrong_bank

    # Emulators commonly boot AOROM, BNROM, GNROM, and
    # UNROM (Crazy Climber) to the first bank.  There's a stub
    # in $BF6C that tries switching to the last bank.
    # Put it in all 16K banks.
    gnromstub = primary_prgrom[0x3F6C:0x3F80]
    for i in range(0, dupli_prgsize, 16384):
        prgrom[i + 0x3F6C:i + 0x3F80] = gnromstub
    for i in range(0, dupli_prgsize - 16384, 16384):
        prgrom[i + 0x3FFC] = 0x6C
        prgrom[i + 0x3FFD] = 0xBF

    # SUROM/SXROM duplication
    prgrom = prgrom * (prgsize // len(prgrom))

    # Place right bank in first 16K (for #180 UNROM Crazy)
    prgrom[:0x3F6C] = primary_prgrom[:0x3F6C]

    # Finally, add bank numbers to PRG ROM
    for i in range(4096 - 8, prgsize, 4096):
        prgrom[i] = i // 4096
        prgrom[i + 1] = 0
    prgrom[-7] = 1

    # Add bank numbers to CHR ROM
    chrrom = bytearray(primary_chrrom) * (chrsize // len(primary_chrrom))
    for i in range(508, chrsize, 1024):
        chrrom[i] = i // 1024

    rom = b''.join([header, prgrom, chrrom])
    return (filename, rom)

def expand_romspec(prgsizes, chrsizes, mapper, mirrors, ramsizes, chrramsizes):
    from collections.abc import Sequence
    from itertools import product

    if not isinstance(prgsizes, Sequence):
        prgsizes = (prgsizes, prgsizes)
    if not isinstance(chrsizes, Sequence):
        chrsizes = (chrsizes, chrsizes)
    if not isinstance(mirrors, Sequence):
        mirrors = (mirrors,)
    if not isinstance(ramsizes, Sequence):
        ramsizes = (ramsizes,)
    if not isinstance(chrramsizes, Sequence):
        chrramsizes = (chrramsizes,)
    ramsizes = (spec
                for sz in ramsizes
                for spec in (set([(sz, 0), (0, sz)])
                             if not isinstance(sz, Sequence)
                             else (sz,)))
    
    return product(log_xrange(*prgsizes),
                   log_xrange(*chrsizes) if chrsizes[0] else (0,),
                   (mapper,), mirrors, ramsizes, chrramsizes)

test_rom_folder = '../testroms'
primary_file = '../mapperel-primary.nes'

def main():
    import os
    from itertools import starmap  # starmap(f, rows = f(*row) for row in rows)
    global primary_prgrom, primary_chrrom
    with open(primary_file, 'rb') as infp:
        infp.read(16)
        primary_prgrom = infp.read(32768)
        primary_chrrom = infp.read(8192)
        
##    h = make_nes2_header(262144, 131072, MAPPER_MMC3, INES_MIRRV,
##                         (0, 8192), 0, INES_NTSC)
##    print(" ".join("%02x" % c for c in h))
    for romspec in romspecs:
        try:
            expand_romspec(*romspec)
        except TypeError:
            print("bad romspec:", romspec)
            raise
    specs = [single_rom
             for romspec in romspecs
             for single_rom in expand_romspec(*romspec)]
    print(len(specs), "specs")
    roms = starmap(handle_single_rom, specs)
    total_sz = 0

    if not os.path.isdir(test_rom_folder):
        os.mkdir(test_rom_folder)
    seen_names = set()
    for (filename, rom) in roms:
        if filename in seen_names:
            print("duplicate filename", filename)
        seen_names.add(filename)
        total_sz += len(rom)
        with open(os.path.join(test_rom_folder, filename), 'wb') as outfp:
            outfp.write(rom)
    print("total: %s" % format_memsize(total_sz))

    outsizes = [2048, 8192, 32768]
    for outsize in outsizes:
        path = os.path.join(test_rom_folder, "%dk.sav" % (outsize // 1024))
        with open(path, "wb") as outfp:
            outfp.write(bytes([0xFF]) * outsize)

if __name__=='__main__':
    main()
